### ISSUE TITLE
Support tagging Kafka events with their schema

### ISSUE THREAD

**Issue by asnare** [divolte/divolte-collector#135]
Divolte currently uses Avro to serialise events as Kafka messages. The [serialisation format](https://avro.apache.org/docs/1.8.1/spec.html#Data+Serialization) includes only the event data and is not self-describing. Deserialisation requires knowledge of the write-schema or the data cannot be interpreted. This poses some challenges for consumers of Divolte events:

 - Consumers must be configured in some way to 'know' the schema used to write the events.
 - Schema evolution is complicated: changes to the schema pose significant deployment challenges.

For Avro files this problem is solved using the [Object Container Format](https://avro.apache.org/docs/1.8.1/spec.html#Object+Container+Files), but the overhead of this is considered too significant to be practical for each message.

One solution to this problem is used by the Confluent platform: schemas are registered in a central registry and each message is prefixed with a header that contains an identifier assigned by the registry. Confluent messages are serialised as:

1. A header, consisting of 5 bytes:

  1. A single byte with a value of 0.
  2. Four bytes containing the ID (an integer) of the schema, in big-endian order.

2. The serialised Avro record.

We're not happy with this solution for the following reasons:

 - The central registry is another component that needs to be deployed and maintained.
 - It introduces a runtime dependency on startup whereby Divolte (or any producer) needs to be able to contact the registry before placing messages on a queue.

With this in mind we would prefer to implement an alternative design that is similar but differs in the header format. Specifically, a message will consist of:

1. A header:

  1. A sequence of 10 fixed bytes: `ff f0 fe f1 f0 d1 c0 de ff ff`
  2. A sequence of 32 bytes a SHA-256 digest of the schema in [Parsing Canonical Form](https://avro.apache.org/docs/1.8.1/spec.html#Parsing+Canonical+Form+for+Schemas)

2. The serialised Avro record.

In the tradition of the specification for Object Container Format, our own Message Container Format header is thus described by the following schema:

```json
{"type": "record", "name": "io.divolte.avro.kafka.message.Header",
 "fields": [
   {"name": "magic",    "type": {"type": "fixed", "name": "Magic",      "size": 10}},
   {"name": "schemaId", "type": {"type": "fixed", "name": "SHA256Hash", "size": 32}}
 ]
}
```

Although this has higher overhead than the Confluent version, I believe that:

 - The overhead should be a small fraction of the size of encoded Divolte events;
 - Kafka's should be able to compress this significantly because it's the same for every message within in a batch.

The benefits of this format are that:

 - Consumers can reliably detect messages encoded using this schema; the magic is very difficult to produce by accident.
 - Producers can start writing messages to a topic immediately without first contacting an external service.
 - Schema evolution is supported on a single topic: consumers can detect which write-schema was used for each message, thereby allowing schema resolution to properly take place.
 - An unaware consumer is unlikely to accidentally misinterpret a message with this header: the magic is designed to trigger errors during deserialisation for many schemas. However clients that are unaware of the header (e.g. Logstash) can be configured to work by modifying the assumed write-schema to have a 42-byte fixed field at the start, thereby skipping the header during serialisation.

The work for this issue involves:

 - Updating the Kafka sink configuration to specify whether this system should be used.
 - Documentation updates.
 - Implementing the encoding. With tests.

Kafka sinks configuration will gain a new property, `container`, which can have the following values:

 - `divolte` (the format described here)
 - `naked` (default, and current behaviour)

References:

 - [Confluent-compatible serialiser](https://github.com/confluentinc/schema-registry/blob/master/avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroSerializer.java#L73)

**Comment by nextdude** [comment]
I agree the hash identifier has benefits over confluent's integer id. But where does one find an existing schema registry product that works with this kind of id? Maybe you guys could support both kinds, so those of us actively looking to operationalize this soon and who have already deployed confluent could make use of it?

**Comment by asnare** [comment]
I agree it's worth supporting both, but was thinking more along the lines of "configure the confluent integer id, and we'll use that" rather than interacting with the registry directly.

Would that suffice?

**Comment by nextdude** [comment]
Yes, I think it would be a valuable increment. I've just started looking at it, but my sense of the code is that, at the moment, divolte's internal schema registry, and the management of schema files, is entirely local and immutable, assumed fixed from the start. To allow schema evolution, and make the life cycle of schema changes driven by changes in an external schema registry seems like a lot of architectural work. I don't know if you have plans for a more dynamic method for deploying updated configurations, but I'd be happy to get involved if you're interested.

One additional step beyond manually configuring a schema's id is to just load the schemas up front by name from the remote registry.  For any schemas so configured, pull their current versions from the registry and populate divolte's internal schema registry and keep track of the version numbers. Then the kafka serializer adds confluent's 5 byte header to the front of flushed kafka messages based on those loaded ids. With this approach, anytime a schema changes in the registry, we just need to restart divolte, and the new definitions are loaded without needing to mess with files and configuring version numbers. The only configuration item is a schema registry url.

I can help with that approach if you think its valuable. In the meantime, manually configured version numbers would be a big win I think.

**Comment by nextdude** [comment]
I just reread this thread and I know you're concerned about divolte depending on contacting another remote service before it can get about its business. And I wouldn't suggest you *require* divolte users to have confluent configured in order to use divolte. But I think there are many companies with experience deploying microservices and who already have confluent running in their ecosystem and would find the ability to support confluent's schema registry in divolte a substantial win.

**Comment by friso** [comment]
One of the goals is indeed to keep the out-of-the-box experience as simple as feasible for Divolte. For this reason it also supports writing to HDFS directly without going through Kafka. Requiring Confluent's components to be present would narrow the target audience.

Right now it's indeed not trivial to make the schema's more dynamic internally, due to the fact that Divolte serialises to bytes early on and only passes along that representation. That said, in the longer run, reloading some parts of the configuration at runtime is certainly something that's been considered. One good reason to support this is that Divolte builds up some internal state for its de-duplication logic, which is lost on a restart. The value of this would depend on how often changes to the config are made, of course. In practice, Divolte has been long running without changes in most of our installations.

Another concerns with dynamically updating configuration and/or externalising some of the configuration (e.g. by requiring a schema registry) is that deployments are not immutable and less self contained. As you mention, this would require the schema registry to be present before Divolte can go about its business. What do you do when it's not there? Or when it's there, but the schema does not match the mapping? Fail on startup, probably. But if you dynamically reload configs and the new config is in error? Normally, you'd tackle this with a rolling deployment which will just fail when the first instance doesn't come up and you stay in the old config state automatically. In this setting it is important that when the first restart succeeds, all subsequent ones will as well, in which case reducing external dependencies is a benefit. So, altogether I think the benefit of dynamic config reloading depends a lot on the deployment setup in use. In my current environment, I favour immutable config, as we deploy using containers and changing a config means building a new container image anyway. Also, deployments create the new instances before shutting down the old ones, so if the deployment fails, nothing is lost. In a more traditional setup (e.g. two or three dedicated machines / VMs for Divolte), I can understand the benefit of being able to reload dynamically, but I'm not sure how common that will still be in the nearby future.

**Comment by fvlankvelt** [comment]
Agreeing with the minimal runtime dependencies, I've created a PR (https://github.com/divolte/divolte-collector/pull/146) that adds none but does enable divolte to write out confluent-compatible messages.

The responsibility of retrieving the schema_id is passed onto the deployment script; so it can decide whether to retrieve it from a service & die if that service is not available, or fetch it from some environment variable, etcetera.

An alternative to dynamic config changes could be to externalize the state that divolte retains.  Write out the state; restart the service with new configuration; restore the state; resume regular operation.  IMO event de-duplication is a 'best-effort' approach anyway, though.  Consumers can be expected to put in some effort themselves if they need strict guarantees.

**Comment by friso** [comment]
Yeah, I've thought about persisting the filter state for de-duping, but in container deployments that's hardly ever helpful, as a new instance won't likely startup with pre-existing storage anyway.

You're right that making the process of fetching the ID from the registry part of the deployment as opposed to part of Divolte's startup is a better separation in terms of failures scenarios.

**Comment by friso** [comment]
Confluent's Schema Registry has a built in safe guard against non-compatible schemas for the same topic / subject. It can be configured to allow either backward only, forward only and full compatibility between existing and new schemas.

Just leaving a note here to document that this exists and we should probably also have it in a meta-topic based solution for schema tagging.

**Comment by asnare** [comment]
I noticed recently that Kafka now supports attaching headers to messages. Instead of inventing a custom container for the message payload as proposed at the top we should simply attach the schema fingerprint as a header. This would:

 - Be similar to what we're already doing when sending messages to Google Pub/Sub.
 - Be much simpler and less confusing than inventing a new custom header.
 - Allow an easier migration path; a topic could be consumed simultaneously by naïve consumers and ones that understand the header. (The schema can't vary until all the naïve consumers have been "upgraded", but the migration path involves less steps.)
 - Allow simultaneous support for the Confluent wire format.

**Comment by asnare** [comment]
I just tested this quickly; the headers are only supported by brokers from Kafka 0.11 and up. On older versions an exception is thrown if the producer sets headers on a message. This means the behaviour will need to be configurable.

**Comment by asnare** [comment]
I spent some time today looking into whether we could do this without a configuration knob by detecting support for headers automatically.

The current situation is:

 - The Kafka producer API doesn't expose this information directly.
 - Support can vary per partition since clusters don't have to be homogenous.
 - We can't easily figure out which partition a record will be assigned to (unless we explicitly assign it ourselves).
 - Support can vary with time as brokers come and go.

   - Detecting this will therefore inherently be racy.

 - The current error is thrown by the producer code that serialises the record into the batch-buffer pending send.

      - When starting a batch, it is marked with the features it supports: this is the basis for the check when appending a record.
      - If the target broker is downgraded between batch start and flush, the buffer is transparently adjusted during send. (Headers are discarded if necessary.)

As things stand, even if we could detect it we wouldn't be robust against cluster changes. The following sequence is the unhappy flow:

1. Check for support. Computer says "yes".
2. Prepare and send the record with the header.
3. Producer notices there is no batch underway, and starts a new one. Since we checked, one of the nodes has downgraded.
4. The new batch-buffer throws an exception because headers aren't supported.
